dojson/contrib/marc21/utils.py

```diff
@@ -24,6 +23,140 @@
 """Location of the MARC21 DTD file"""
 
 
+class GroupableOrderedDict(dict):
+    """Immutable list that can group values pretending to be a dict."""
+
+    def __new__(cls, values=None):
+        new = dict.__new__(cls)
+        dict.__init__(new)
+
+        data = []
+        keys = []
+        ordering = []
+
+        if values:
+            if isinstance(values, GroupableOrderedDict):
+                data = values._data
+                keys = values._keys
+                ordering = values.__order__
+            elif isinstance(values, dict):
+                if "__order__" not in values:
+                    raise AttributeError("Can only be recreated from an "
+                                         "ordered dict.")
+
+                ordering = values["__order__"]
+                counter = Counter()
+                for key in ordering:
+                    if key not in keys:
+                        keys.append(key)
+                    value = values[key][counter[key]]
+                    data.append((key, value))
+                    counter[key] += 1
+            else:
+                for key, value in values:
+                    if key not in keys:
+                        keys.append(key)
+                    data.append((key, value))
+                    ordering.append(key)
+
+        # immutable, all the things
+        new._data = tuple(data)
+        new._keys = tuple(keys)
+        new.__order__ = tuple(ordering)
+        return new
+
+    def __init__(self, values=None):
+        pass
+
+    def __copy__(self):
+        return GroupableOrderedDict(self.items(repeated=True))
+
+    def __deepcopy__(self):
+        return self.__copy__()
+
+    def __reduce__(self):
+        return GroupableOrderedDict, (dict(self.items()), )
+
+    def __eq__(self, other):
+        if len(other) != len(self._keys):
+            return False
+
+        for k, v in self.iteritems(with_order=False):
+            if len(v) != len(other[k]):
+                return False
+
+            for i, value in enumerate(v):
+                if other[k][i] != value:
+                    return False
+        return True
+
+    def get(self, key, default=None):
```

Comment:
It doesn't seem to work. This is a `dict` and not a `Mapping` or `MutableMapping` anymore.